### ISSUE TITLE
EdgesGeometry shows unnecessary edges

### ISSUE THREAD

**Issue by Emanuele-Spatola** [mrdoob/three.js#10517]
##### Description of the problem 
The EdgesGeometry takes all the cords shared between faces and if the faces are not coplanar (or their "dot" is less than the thresholdDot) then it shows them as edges.

The problem is that in some cases a cord is the sum of two other cords in two different faces. In this case the cord is shown as edge even if its's not.

This is the wireframe:
![screen shot 2017-01-06 at 12 02 01](https://cloud.githubusercontent.com/assets/5675600/21715998/01b68580-d408-11e6-911c-d43090dbc16d.png)

and this is with the EdgesGeometry:
![screen shot 2017-01-06 at 12 01 44](https://cloud.githubusercontent.com/assets/5675600/21715997/01b2fcf8-d408-11e6-9175-f8628cd184c6.png)

I've made a plunker with an example geometry:
https://plnkr.co/edit/43Sd9vsiAUsB7pTCUMEQ

I'm thinking we should, for each cord, check if there is a vertex in it, and if so, split the face in two using that vertex.
Any idea on how we could do it?

##### Three.js version

- [x] Dev
- [x] r82
- [ ] ...

##### Browser

- [x] All of them
- [ ] Chrome
- [ ] Firefox
- [ ] Internet Explorer

##### OS

- [x] All of them
- [ ] Windows
- [ ] Linux
- [ ] Android
- [ ] IOS

**Comment by Emanuele-Spatola** [comment]
@WestLangley you may be very helpful with this :)

**Comment by WestLangley** [comment]
>The EdgesGeometry takes all the cords shared between faces and if the faces are not coplanar (or their "dot" is less than the thresholdDot) then it shows them as edges.

Right. But you have a long edge that is coincident with edges of two smaller triangles --_coincident_, but not _shared_.

That means, the long edge is not shared at all. Nor are the two short edges.

Edges that are not shared are rendered by `EdgesGeometry`.

**Comment by Emanuele-Spatola** [comment]
> Right. But you have a long edge that is coincident with edges of two smaller triangles --coincident, but not shared.

Yes, that's why I suggested to split the triangle in two using the vertex of the coincident edge.

I updated the plunker adding `EdgesGeometry2` that does that:
https://plnkr.co/edit/43Sd9vsiAUsB7pTCUMEQ

I'm sure the code can be optimised, but I think it will always be quite slow for complex geometries.

**Comment by WestLangley** [comment]
>I'm thinking we should, for each cord, check if there is a vertex in it, and if so, split the face in two using that vertex.

Sorry, I do not think it is the responsibility of `EdgesGeometry` to split faces in your geometry.

I also think that `EdgesGeometry` is doing exactly what it is supposed to do. I respect your alternate view, however.

**Comment by Emanuele-Spatola** [comment]
Well I think if a cord is not an edge it should't be shown.

I understand that splitting polygons may seem too much for the `EdgeGeometry`, but I also think it's wrong to split the geometry somewhere else to "make it good" for `EdgeGeometry`.

Maybe the the geometry I used is a particular case (it's the result of CSG operations), but it's a possible case and unless we say it's somehow an invalid geometry it should be handled correctly.

**Comment by WestLangley** [comment]
>Well I think if a cord is not an edge it should't be shown.

You are asking `EdgesGeometry` to render part of a triangle's edge.

>unless we say it's somehow an invalid geometry it should be handled correctly.

I, too, do not think your geometry is "invalid". Maybe you can come up with an efficient algorithm to handle this use case.

**Comment by Wilt** [comment]
@Emanuele-Spatola Could you share this EdgesGeometry2 you are mentioning? 
It seems not to be inside the plnkr that you shared.

**Comment by Emanuele-Spatola** [comment]
@Wilt you are right, I shared the wrong link.
I created a new plunker, and called the new geometry as you suggested:
https://plnkr.co/edit/uEYo6L3pgbIaYXXzVzXd

I'd be happy to tidy the code up and create a pull request if the code is useful to other people.
My only concern is about the performance of the code, as it takes O(edges*faces) to execute.

@WestLangley what do you think?

**Comment by Wilt** [comment]
@Emanuele-Spatola Thanks for sharing. I don't have time right now, but I will soon have a look at your code to see if I can come up with some possible optimizations.

**Comment by FishOrBear** [comment]
![image](https://user-images.githubusercontent.com/19372111/29764503-eecb26e6-8c09-11e7-9bbb-9614a3e934a8.png)
![image](https://user-images.githubusercontent.com/19372111/29764508-f2d32fea-8c09-11e7-875e-dccab668b8d6.png)
I think your proposal is useful.

**Comment by stephan157** [comment]
@Emanuele-Spatola
Is there also a version of the OutlinesGeometry which works with the current version of THREE.js?
As I don't have that deep knowledge of THREE.js I don't know how the code has to be

**Comment by Mugen87** [comment]
Agreeing with https://github.com/mrdoob/three.js/issues/10517#issuecomment-271355253. `EdgesGeometry` is not an outline effect.
